### ISSUE TITLE
Inject dependencies and set initial values for runtime data

### ISSUE THREAD

**Issue by payam-int** [xaviergonz/mobx-keystone#284]
Hi,
Thanks for your amazing work!

I want to inject dependencies or set initial values for [runtime data](https://mobx-keystone.js.org/classModels#runtime-data) and have access to them at `onInit` but I don't if it is possible.

For now my solution was using `onAttachedToRootStore` and context.
```js
class MyModel extends Model(...){}
const DependenciesContext = createContext()
const createStore = (initialValues, dependencies){
    const model = new MyModel(initialValues);
    DependenciesContext.set(model, dependencies);
    registerRootStore(model);
    return model;
}
```
But it looks really hacky. It would be great if I could have an API like [MST dependency injection](https://mobx-state-tree.js.org/concepts/dependency-injection).

Thanks

**Comment by xaviergonz** [comment]
Note that in MST it is also true that you cannot access the parent env (context) until the node is attached to the parent:

```js
import * as mst from "mobx-state-tree"

const M = mst.types.model({})
.views((self) => {
  return {
    getEnv() {
      return mst.getEnv(self)
    }
  }
})

const R = mst.types.model({
  m: mst.types.maybe(M)
})
.actions(self => {
  return {
    setM(m) {
      self.m = m
    }
  }
})

const r = R.create({}, {some: "env"})
const m = M.create({})
console.log(m.getEnv()) // prints {}
r.setM(m)
console.log(m.getEnv()) // prints {some: "env"}
```

What if you could pass a function to "new Model" / fromSnapshot that would get invoked before onInit is invoked to allow you to init contexts? would that work?
e.g.
```
const DependenciesContext = createContext()

const model = new MyModel(initialValues, {
  preInit(node) => {
    DependenciesContext.set(node, "whatever value")
    node.setMyVolatileStuff("x")
  }
})

const model = fromSnapshot(snapshot, {
  preInit(node) {
    DependenciesContext.set(node, "whatever value")
    node.setMyVolatileStuff("x")
  }
}
```

would that suffice?

**Comment by xaviergonz** [comment]
or an alternative could be this
```js
const DependenciesContext = createContext()

class MyModel ... {
  onInit(values) {
    DependenciesContext.set(this, values.a)
    this.setSomeVolatile(values.a)
  }
}

const model = new MyModel(initialValues, {
  onInitValues: {
    a: "whatver"
  }
})

const model = fromSnapshot(snapshot, {
  onInitValues: {
    a: "whatver"
  }
})
```

**Comment by xaviergonz** [comment]
Ok, probably this is a better idea to "emulate" envs in mst while relying on contexts

```js
// we have a default context called envContext (defaults to undefined)

const x = new M(..., { env: valueToSetTheContextTo})

// the context will be set on the root before onInit
// to get the value:
envContext.get(node)

fromSnapshot(..., { env: valueToSetTheContextTo })
```

**Comment by sisp** [comment]
But `envContext` would not be typed well, would it? Wed would need to rely on a type assertion or type guard.

An explicit context defined by the user could be typed more strongly.

**Comment by xaviergonz** [comment]
> But envContext would not be typed well, would it? We would need to rely on a type assertion or type guard.

Good point.

back to preInit then?

```js
const envContext = createContext()

const x = new M(..., { // or fromSnapshot
  preInit: n => envContext.set(n, value)
})
```

**Comment by xaviergonz** [comment]
Ok, thinking about it some more, preInit might be too unrestrictive. Since it would need to be run right after the node is created, but no children/props are attached to it, once might be tempted to access children in the preinit phase and would be confused why there are none. I think restricting it to context initialization might be better:

```js
const envContext = createContext()

const x = new M(..., { // or fromSnapshot, or clone, or toTreeNode
  contexts: [
    envContext.withValue(X)
  ]
  // alternatively, if one context
  context: envContext.withValue(X)
})
```

by doing that we would be able to ensure all children (be it on onInit, onAttachedToRootStore or whichever phase) have access to the root context

**Comment by sisp** [comment]
A general question: In most cases, I assume the env context to be an object of key-value pairs. What's the behavior if a non-root node is (pre-)initialized with a partial env context? Should the env context of this node be merged with the parent env context and the keys specified in this node's env context take precedence? With an approach like passing `envContext.withValue(...)` to the constructor, at least right now the context would be replaced. Perhaps (even independent of the env context feature) contexts should have a way to configure what happens if a parent node already provides the same context, e.g.:
```ts
const envContext = createContext(..., {
  merge: (current, parent) => ({ ...parent, ...current })
})
```
The default behavior would be:
```ts
createContext(..., {
  merge: (current, parent) => current
})
```

**Comment by xaviergonz** [comment]
The current behavior is replacing, but it could be configurable like you pointed out. Do you have any use cases in mind?

**Comment by sisp** [comment]
Off the top of my head:
1. Partially provided env contexts at different nodes in the tree (for whatever reason, but maybe that's not a use case?).
2. Implementing a custom path (according to domain-specific rules) in the tree using a context, i.e. some nodes would provide the context with their path item (single element array) and the merge rule concatenates it with the parent's path array.

**Comment by xaviergonz** [comment]
Isn't that right now possible (while using onInit)?:
```
subnode onInit() {
  ctx.setComputed(this, () => [...ctx.get(getParent(this)), "subnodePath"]);
}
```

In the end I'm thinking of using something like this

```
const envContext = createContext(...)

envContext.apply(() => new X(), {...})
envContext.apply(() => fromSnapshot(...), {...})
envContext.apply(() => toTreeNode(...), {...})
```

Basically what it would do is:
- save the current context default value
- set the default value to the provided one (well, actually a new value called temp override)
- run the code
  - this means that every node being created inside the function will be able to access the value in onInit
- if the function returns a node, then the context will be set to that value for that particular node (so that any future operations on that root node / subnodes can still access that context value)
- the old default will be restored

I was hitting a wall with fromSnapshot, where I needed subobjects (leafs) to get be able to get the context value while their parent's didn't even get initialized yet)

**Comment by sisp** [comment]
> Isn't that right now possible (while using onInit)?:
> ```
> subnode onInit() {
>   ctx.setComputed(this, () => [...ctx.get(getParent(this)), "subnodePath"]);
> }
> ```

Yes, that's actually how I've been doing it so far. The `merge` option in `createContext` would move the logic from a model to the context though, might be favorable.

The `envContext.apply(...)`  API looks nice, also how it would be used like you showed in https://github.com/xaviergonz/mobx-keystone/issues/284#issuecomment-902707182. 🙂 Perhaps just `contexts: [...]` and no `context` to keep it simple?

**Comment by xaviergonz** [comment]
I was actually thinking if it is worth it. As is `apply` would work for every current and future creation-related API, while if I add it to `new` then I also need to add it to fromSnapshot, toTreeNode, clone, and whatever might come later.

**Comment by sisp** [comment]
Ah, you're thinking to use `apply` only without changing `new` etc.? Sounds good to me.

**Comment by xaviergonz** [comment]
@payam-int any thoughts?

**Comment by payam-int** [comment]
> @payam-int any thoughts?

It's all good!

**Comment by justin-hackin** [comment]
Not much useful to contribute here but just wanted to say I really appreciate these considerations. I've been thinking about ways in which to define nested models in MST so that the model's properties contain metadata about how the property should be controlled e.g. a slider with min=0, max=1, step=0.1, or a color picker. In this way I could, for example, render a control panel for the model in different component frameworks (or pure HTML5 inputs) without redefining the data that is associated with the view layer. Given a model, I could traverse the tree and automatically generate a control panel while still allowing custom controls for particular properties.  This discussion https://github.com/mobxjs/mobx-state-tree/discussions/1758 led me to consider if I could solution this with mobx-keystone. With a context as pseudo-env feature it seems I could so I'm considering converting to mobx-keystone. My use case is to create a general purpose plugin-based framework for creating SVG generation widgets out of the design-specific app I'm building here https://www.youtube.com/watch?v=_J6QBTSfmpU

**Comment by xaviergonz** [comment]
context apply/applyComputed was released as part of v0.61.0
docs: https://mobx-keystone.js.org/contexts (might take a few mins to update)
